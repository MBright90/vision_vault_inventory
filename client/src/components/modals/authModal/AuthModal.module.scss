@import '~@style/_variables.scss';

.modalBackground {
    background-color: rgba(100, 100, 100, 0.8);
    bottom: 0;
    display: grid;
    left: 0;
    position: absolute;
    place-content: center;
    top: 0;
    right: 0;
}

.modalContainer {
    align-items: center;
    background-color: $dark-red;
    color: $light-cream;
    display: flex;
    flex-direction: column;
    gap: 15px;
    padding: 30px;
    position: relative;

    .closeModal {
        background: none;
        color: $light-cream;
        height: 20px;
        padding: 3px !important;
        position: absolute;
        right: 5px;
        top: 5px;
        width: 20px !important;
    }

    input {
        width: 96%;
    }
    
    .buttonsContainer {
        display: flex;
        gap: 5%;
        justify-content: center;

        .confirmButton {
            margin-top: 15px;
            padding: 8px 30px;
            width: fit-content;
        }
    }
}