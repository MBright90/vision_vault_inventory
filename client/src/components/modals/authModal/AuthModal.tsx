import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faXmark } from "@fortawesome/free-solid-svg-icons";

import style from './AuthModal.module.scss';

interface AuthModalProps {
    closeModal: () => void,
    callback: (() => void) | (() => Promise<void>)
}

const AuthModal: React.FC<AuthModalProps> = ({ closeModal, callback }) => {
    return (
        <div className={style.modalBackground}>
            <div className={style.modalContainer}>
                <button className={style.closeModal} onClick={closeModal}><FontAwesomeIcon icon={faXmark}/></button>
                <p>Enter admin password to confirm:</p>
                <input className={style.passwordInput}/>
                <button className={style.confirmButton} onClick={() => { void callback(); closeModal(); }}>
                    Confirm
                </button>
            </div>
        </div>
    );
};

module.exports = AuthModal;