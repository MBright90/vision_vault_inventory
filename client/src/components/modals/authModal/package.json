{
    "name": "auth-modal",
    "version": "0.0.0",
    "main": "AuthModal.tsx",
    "dependencies": {
        "react": "^18.2.0"
    }
}
