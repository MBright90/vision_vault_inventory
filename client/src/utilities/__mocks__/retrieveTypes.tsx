import React from "react";

export default jest.fn(async (setFilterOptions, setFilterIdArr, selected) => {
    const data = [
        { _id: 'type1', name: 'Type 1' },
        { _id: 'type2', name: 'Type 2' },
    ];

    const filterNodes = data.map(type => (
        <option key={type._id} data-id={type._id} selected={selected === type.name}>
            {type.name}
        </option>
    ));
    setFilterOptions(filterNodes);

    if (setFilterIdArr) {
        const filterIds = data.map(type => ({ name: type.name, _id: type._id }));
        setFilterIdArr([{ name: 'all', _id: 'all' }, ...filterIds]);
    }
});