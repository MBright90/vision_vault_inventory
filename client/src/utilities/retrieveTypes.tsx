import React from "react";
import capitalize from "./capitalize";

import { type productType_type } from "@custom_types/types";
import endpoint from "./endpoint";

interface FilterIdSet{
    name: string,
    _id: string
}

const retrieveTypes = async (
    setFilterOptions: (nodes: React.ReactNode[]) => void,
    setFilterIdArr: ((idSet: FilterIdSet[]) => void) | null,
    selected: string | null = null,
    ): Promise<void> => {
        let formattedSelected: string = "";
        if (selected !== null) {
            formattedSelected = capitalize(selected);
        }
    let data: productType_type[] = [];
    try {
        const response = await fetch(`${endpoint}/types/`);
        data = await response.json();
        // Create ReactNode arr
        const filterNodes: React.ReactNode[] = data.map((type: productType_type) => {
            if (capitalize(formattedSelected) === type.name) {
                return (<option key={type._id} data-id={type._id} selected={true}>
                    {capitalize(type.name)}
                </option>);
            } else {
                return (<option key={type._id} data-id={type._id}>
                    {capitalize(type.name)}
                </option>);
            }
        });
        setFilterOptions(filterNodes);

        // Create filterId arr
        if (setFilterIdArr !== null) {
            const filterIds: FilterIdSet[] = data.map((type: productType_type) => {
                return { name: type.name, _id: type._id };
            });
            setFilterIdArr([{ name: 'all', _id: 'all' }, ...filterIds]);
        }
    } catch (err) {
        console.log(err);
    }
};

export default retrieveTypes;
