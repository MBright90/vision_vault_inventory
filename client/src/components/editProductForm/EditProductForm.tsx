import React, { useEffect, useState } from "react";

import style from './EditProductForm.module.scss';
import { redirect, useParams } from "react-router-dom";
import retrieveTypes from "@utilities/retrieveTypes";
import endpoint from "@utilities/endpoint";
import Modal from "@components/modal/Modal";
import { type product_type } from "@custom_types/types";

const EditProductForm: React.FC = () => {
    const { productId } = useParams();

    const [modal, setModal] = useState<React.ReactNode | null>(null);

    const [typeOptions, setTypeOptions] = useState<React.ReactNode[]>([]);

    const [currentName, setCurrentName] = useState<string>('');
    const [currentDescription, setCurrentDescription] = useState<string>('');
    const [currentPrice, setCurrentPrice] = useState<number>(0);
    const [currentStock, setCurrentStock] = useState<number>(0);
    const [currentType, setCurrentType] = useState<string>('');
    const [currentGenres, setCurrentGenres] = useState<string>('');

    useEffect(() => {
        const retrieveProduct = async (): Promise<void> => {
            try {
                const response = await fetch(`${endpoint}/products/${productId}`);
                if (!response.ok) {
                    setModal(
                        <Modal closeModal={() => {setModal(null);} }>
                            <p>Error retrieving product</p>
                        </Modal>
                    );
                    return;
                }

                const product: product_type = await response.json();
                setCurrentName(product.name);
                setCurrentDescription(product.description);
                setCurrentPrice(product.price);
                setCurrentStock(product.number_in_stock);
                setCurrentType(product.type.name);
                setCurrentGenres(product.genres.join(', '));
            } catch (err) {
                console.log(err);
            }
        };

        console.log('retrieving');

        // Set options for types
        void retrieveProduct();
        void retrieveTypes(setTypeOptions, null);
    }, []);


    function handlePriceChange(e: React.ChangeEvent<HTMLInputElement>):void {
        const { value } = e.target;

        try {
            // convert value to float, reduce to 2DP, the convert to float again
            const priceValue = parseFloat(value);
            if (priceValue >= 0) setCurrentPrice(parseFloat(priceValue.toFixed(2)));
        } catch (err) {
            setCurrentPrice(0);
            console.log(err);
        }
    }

    function handleStockChange(e: React.ChangeEvent<HTMLInputElement>): void {
        const { value } = e.target;
        
        try {
            const stockValue = parseInt(value);
            if (stockValue >= 0) setCurrentStock(stockValue);
        } catch (err) {
            setCurrentStock(0);
            console.log(err);
        }
    }

    async function submitForm(e: React.MouseEvent<HTMLButtonElement>): Promise<void> {
        e.preventDefault();

        const requestOptions = {
            method: 'PUT',
            headers: { 'Content-Type': 'application/json' }, 
            body: JSON.stringify({
                name: currentName,
                description: currentDescription,
                price: currentPrice,
                number_in_stock: currentStock,
                genres: currentGenres,
                type: currentType
            })
        };

        try {
            const response = await fetch(`${endpoint}/products/edit/${productId}`, requestOptions);
            if (!response.ok) {
                const data = await response.json();
                setModal(
                    <Modal closeModal={() => { setModal(null); }}>
                        <p>Error updating product</p>
                        <p>Error: {data.err}</p>
                    </Modal>
                );
            } else {
                const confirm = { text: 'Ok', func: () => { redirect('/'); } };
                setModal(
                    <Modal closeModal={() => { setModal(null); }} confirm={confirm}>
                        <p>Product edited successfully</p>
                    </Modal>
                );
            }
        } catch (err) {
            console.log(err);
        }
    }

    return (
        <main className={style.formContainer}>
            { modal }
            <h3>Edit Product</h3>
            <form action="" className={style.editProductForm}>

                <div className={style.inputContainer}>
                    <label htmlFor="name-input">Product Name</label>
                    <input id="name-input" name="name-input" type="text" value={currentName} onChange={(e) => { setCurrentName(e.target.value); }} required />
                </div>

                <div className={style.inputContainer}>
                    <label htmlFor="description-input">Description</label>
                    <textarea id="description-input" name="description-input" value={currentDescription} onChange={(e) => { setCurrentDescription(e.target.value); }} required />
                </div>

                <div className={style.inputContainer}>
                    <label htmlFor="price-input">Price</label>
                    <input id="price-input" name="price-input" type="number" min={0} value={currentPrice} step={.01} onChange={(e) => { handlePriceChange(e); }} required />
                </div>

                <div className={style.inputContainer}>
                    <label htmlFor="stock-input">Current stock</label>
                    <input id="stock-input" name="stock-input" type="number" max={100} min={0} value={currentStock} step={1} onChange={(e) => { handleStockChange(e); }} required />
                    <p className={style.inputInfo}>Max 100</p>
                </div>

                <div className={style.inputContainer}>
                    <label htmlFor="type-input">Type</label>
                    <select id="type-input" name="type-input" value={currentType} onChange={(e) => { setCurrentType(e.target.value); }} required>
                        <option>Select Type...</option>
                        {typeOptions}
                    </select>
                </div>

                <div className={style.inputContainer}>
                    <label htmlFor="genres-input">Genres</label>
                    <input id="genres-input" name="genres-input" type="text" value={currentGenres} onChange={(e) => { setCurrentGenres(e.target.value); }} required />
                    <p className={style.inputInfo}>Minimum of one required</p>
                    <p className={style.inputInfo}>Separate genres with a comma</p>
                </div>

                <button className={style.formSubmitBtn} type="submit" onClick={(e) => { void submitForm(e); }}>Update Product</button>
            </form>
        </main>
    );
};

export default EditProductForm;
