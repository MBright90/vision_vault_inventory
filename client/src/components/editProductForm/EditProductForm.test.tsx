import React from "react";
import { screen, render, fireEvent, waitFor } from "@testing-library/react"
import "@testing-library/jest-dom"
import EditProductForm from "./EditProductForm"

import { MemoryRouter, Route, Routes } from "react-router-dom";
import { ReqOptions } from "@components/modals/authModal/AuthModal";
import retrieveTypes from "@utilities/retrieveTypes";

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useParams: jest.fn(),
}));

jest.mock('@components/modals/confirmModal/ConfirmModal', () => ({
    __esModule: true,
    default: (props: { children: React.ReactNode[], closeModal: () => {} }) => {
        return (
            <div data-testid="mock-confirm-modal">
                {props.children}
            </div>
        )
    }
}));

jest.mock('@components/modals/authModal/AuthModal', () => ({
    __esModule: true,
    default: (props: { closeModal: () => {}, reqOptions: ReqOptions, productId: string, displayModalFunc: (message: string) => {} }) => {
        return (
            <div data-testid="mock-auth-modal">
                {props.reqOptions.body}
            </div>
        )
    }
}));

jest.mock('@utilities/retrieveTypes', () => jest.fn());

beforeEach(() => {
    jest.resetAllMocks();

    global.fetch = jest.fn((): Promise<Response> => 
        Promise.resolve({
            ok: true,
            json: async () => ({
                _id: 'test-id',
                name: 'test-name',
                description: 'test-description',
                price: 1,
                number_in_stock: 2,
                type: { name: 'test-type-name' },
                genres: [
                    { name: 'test-genre-1', _id: 'test-genre-id-1' },
                    { name: 'test-genre-2', _id: 'test-genre-id-2' },
                ],
            }),
        } as Response)
    );
});


test('matches snapshot', async () => {
    (require('react-router-dom').useParams as jest.Mock).mockReturnValue({ productId: 'test-id' });

    const { container } = render(
        <MemoryRouter initialEntries={['/edit/test-id']}>
            <Routes>
                <Route path="/edit/:productId" element={<EditProductForm />}>
                </Route>
            </Routes>
        </MemoryRouter>
    );

    await waitFor(() => {
        expect(container).toMatchSnapshot();
    });
});

test('calls retrieveTypes on component mount', async () => {
    (require('react-router-dom').useParams as jest.Mock).mockReturnValue({ productId: 'test-id' });

    render(
        <MemoryRouter initialEntries={['/edit/test-id']}>
            <Routes>
                <Route path="/edit/:productId" element={<EditProductForm />}>
                </Route>
            </Routes>
        </MemoryRouter>
    );

    await waitFor(() => {
        expect(retrieveTypes).toHaveBeenCalledTimes(1);
    });
});

test('renders form with product info values in inputs', async () => {
    (require('react-router-dom').useParams as jest.Mock).mockReturnValue({ productId: 'test-id' });

    render(
        <MemoryRouter initialEntries={['/edit/test-id']}>
            <Routes>
                <Route path="/edit/:productId" element={<EditProductForm />}>
                </Route>
            </Routes>
        </MemoryRouter>
    );

    await waitFor(() => {
        expect(screen.getByLabelText(/Product Name/i)).toHaveValue('test-name');
        expect(screen.getByLabelText(/Description/i)).toHaveValue('test-description');
        expect(screen.getByLabelText(/Price/i)).toHaveValue(1);
        expect(screen.getByLabelText(/Current stock/i)).toHaveValue(2);
        expect(screen.getByLabelText(/Type/i)).toHaveValue('Select Type...');
        expect(screen.getByLabelText(/Genres/i)).toHaveValue('test-genre-1, test-genre-2');
    });
});

test('updates the product name input value when typed into', async () => {
    (require('react-router-dom').useParams as jest.Mock).mockReturnValue({ productId: 'test-id' });

    render(
        <MemoryRouter initialEntries={['/edit/test-id']}>
            <Routes>
                <Route path="/edit/:productId" element={<EditProductForm />}>
                </Route>
            </Routes>
        </MemoryRouter>
    );
    
    const nameInput = screen.getByLabelText(/Product Name/i);
    fireEvent.change(nameInput, { target: { value: 'Test Product Name' } });
    await waitFor(() => {
        expect(nameInput).toHaveValue('Test Product Name');
    });
});

test('updates the product description input value when typed into', async () => {
    (require('react-router-dom').useParams as jest.Mock).mockReturnValue({ productId: 'test-id' });

    render(
        <MemoryRouter initialEntries={['/edit/test-id']}>
            <Routes>
                <Route path="/edit/:productId" element={<EditProductForm />}>
                </Route>
            </Routes>
        </MemoryRouter>
    );

    const descriptionInput = screen.getByLabelText(/Description/i);
    fireEvent.change(descriptionInput, { target: { value: 'Test Product Description' } });
    await waitFor(() => {
        expect(descriptionInput).toHaveValue('Test Product Description');
    });
});

test('updates the product price input value when typed into', async () => {
    (require('react-router-dom').useParams as jest.Mock).mockReturnValue({ productId: 'test-id' });

    render(
        <MemoryRouter initialEntries={['/edit/test-id']}>
            <Routes>
                <Route path="/edit/:productId" element={<EditProductForm />}>
                </Route>
            </Routes>
        </MemoryRouter>
    );

    const priceInput = screen.getByLabelText(/Price/i);
    fireEvent.change(priceInput, { target: { value: '100' } });
    await waitFor(() => {
        expect(priceInput).toHaveValue(100);
    });
});

test('updates the product initial stock input value when typed into', async () => {
    (require('react-router-dom').useParams as jest.Mock).mockReturnValue({ productId: 'test-id' });

    render(
        <MemoryRouter initialEntries={['/edit/test-id']}>
            <Routes>
                <Route path="/edit/:productId" element={<EditProductForm />}>
                </Route>
            </Routes>
        </MemoryRouter>
    );

    const stockInput = screen.getByLabelText(/Price/i);
    fireEvent.change(stockInput, { target: { value: '200' } });
    await waitFor(() => {
        expect(stockInput).toHaveValue(200);
    });
});

test('updates the product genres input value when typed into', async () => {
    (require('react-router-dom').useParams as jest.Mock).mockReturnValue({ productId: 'test-id' });

    render(
        <MemoryRouter initialEntries={['/edit/test-id']}>
            <Routes>
                <Route path="/edit/:productId" element={<EditProductForm />}>
                </Route>
            </Routes>
        </MemoryRouter>
    );

    const genresInput = screen.getByLabelText(/Genres/i);
    fireEvent.change(genresInput, { target: { value: 'Test Genre One, Test Genre Two'} });
    await waitFor(() => {
        expect(genresInput).toHaveValue('Test Genre One, Test Genre Two');
    });
});

test('AuthModal is displayed when form submitted with valid test-id', async () => {
    (require('react-router-dom').useParams as jest.Mock).mockReturnValue({ productId: 'test-id' });

    render(
        <MemoryRouter initialEntries={['/edit/test-id']}>
            <Routes>
                <Route path="/edit/:productId" element={<EditProductForm />}>
                </Route>
            </Routes>
        </MemoryRouter>
    );

    const submitBtn = screen.getByText(/^Update product/i);
    fireEvent.click(submitBtn);

    await waitFor(() => {
        expect(screen.getByTestId('mock-auth-modal')).toBeInTheDocument();
    });
});

test('ConfirmModal is displayed when form submitted with undefined productId', async () => {
    (require('react-router-dom').useParams as jest.Mock).mockReturnValue({ productId: undefined });

    render(
        <MemoryRouter initialEntries={['/edit/test-id']}>
            <Routes>
                <Route path="/edit/:productId" element={<EditProductForm />}>
                </Route>
            </Routes>
        </MemoryRouter>
    );

    const submitBtn = screen.getByText(/^Update product/i);
    fireEvent.click(submitBtn);

    await waitFor(() => {
        expect(screen.getByTestId('mock-confirm-modal')).toBeInTheDocument();
    });
});

test('ConfirmModal is displayed if error retrieving product on component mount', async () => {
    (require('react-router-dom').useParams as jest.Mock).mockReturnValue({ productId: 'test-id' });

    global.fetch = jest.fn((): Promise<Response> =>
        Promise.resolve({
            ok: false,
        } as Response)
    );

    render(
        <MemoryRouter initialEntries={['/edit/test-id']}>
            <Routes>
                <Route path="/edit/:productId" element={<EditProductForm />}>
                </Route>
            </Routes>
        </MemoryRouter>
    );

    const submitBtn = screen.getByText(/^Update product/i);
    fireEvent.click(submitBtn);

    await waitFor(() => {
        expect(screen.getByTestId('mock-confirm-modal')).toBeInTheDocument();
        expect(screen.getByText(/Error retrieving product/i)).toBeInTheDocument();
    });
});


test('handles component unmounting without errors', () => {
    (require('react-router-dom').useParams as jest.Mock).mockReturnValue({ productId: 'test-id' });

    const { unmount } = render(
        <MemoryRouter initialEntries={['/edit/test-id']}>
            <Routes>
                <Route path="/edit/:productId" element={<EditProductForm />}>
                </Route>
            </Routes>
        </MemoryRouter>
    );

    expect(() => unmount()).not.toThrow();
});