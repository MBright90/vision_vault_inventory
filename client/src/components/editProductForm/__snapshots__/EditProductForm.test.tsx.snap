// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <main>
    <h3>
      Edit Product
    </h3>
    <form
      action=""
    >
      <div>
        <label
          for="name-input"
        >
          Product Name
        </label>
        <input
          id="name-input"
          name="name-input"
          required=""
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          for="description-input"
        >
          Description
        </label>
        <textarea
          id="description-input"
          name="description-input"
          required=""
        />
      </div>
      <div>
        <label
          for="price-input"
        >
          Price
        </label>
        <input
          id="price-input"
          min="0"
          name="price-input"
          required=""
          step="0.01"
          type="number"
          value="0"
        />
      </div>
      <div>
        <label
          for="stock-input"
        >
          Current stock
        </label>
        <input
          id="stock-input"
          max="100"
          min="0"
          name="stock-input"
          required=""
          step="1"
          type="number"
          value="0"
        />
        <p>
          Max 100
        </p>
      </div>
      <div>
        <label
          for="type-input"
        >
          Type
        </label>
        <select
          id="type-input"
          name="type-input"
          required=""
        >
          <option>
            Select Type...
          </option>
        </select>
      </div>
      <div>
        <label
          for="genres-input"
        >
          Genres
        </label>
        <input
          id="genres-input"
          name="genres-input"
          required=""
          type="text"
          value=""
        />
        <p>
          Minimum of one required
        </p>
        <p>
          Separate genres with a comma
        </p>
      </div>
      <button
        type="submit"
      >
        Update Product
      </button>
    </form>
  </main>
</div>
`;
